components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt

```diff
@@ -607,34 +607,45 @@ class QrFragment : Fragment() {
     }
 
     @VisibleForTesting
-    internal fun tryScanningBitmap(bitmap: BinaryBitmap) {
+    internal fun tryScanningSource(source: LuminanceSource) {
         scanCompleteListener?.let {
-            decodeBitmap(bitmap)?.let {
+            decodeSource(source)?.let {
                 scanCompleteListener?.onScanComplete(it)
             }
         }
     }
 
     @VisibleForTesting
-    internal fun decodeBitmap(bitmap: BinaryBitmap): String? {
+    internal fun decodeSource(source: LuminanceSource): String? {
         if (qrState != STATE_DECODE_PROGRESS) {
             return null
         }
 
-        return try {
-            val rawResult = multiFormatReader.decodeWithState(bitmap)
-            if (rawResult != null) {
-                qrState = STATE_QRCODE_EXIST
-                rawResult.toString()
-            } else {
+        val bitmap = createBinaryBitmap(source)
+        val rawResult = try {
+            multiFormatReader.decodeWithState(bitmap)
+        } catch (e: NotFoundException) {
+            val invertedBitmap = createBinaryBitmap(source.invert())
```

Comment:
Right now this only tries the inverted image if a `NotFoundException` was thrown...but looking through various scanner implementations it seems possible that `decodeWithState` just returns a `null` result or throws something else entirely. I also managed to get it to crash (with an `ArrayIndexOutOfBoundsException`) by scanning repeatedly.

What do you think about this small to change to the PR which will handle all these cases? Essentially we just handle the inversion on method further up the stack.

```kotlin
 @VisibleForTesting
    internal fun tryScanningSource(source: LuminanceSource) {
        if (qrState != STATE_DECODE_PROGRESS) {
            return
        }

        scanCompleteListener?.let {
            val result = decodeSource(source) ?: decodeSource(source.invert())
            result?.let {
                scanCompleteListener?.onScanComplete(it)
            }
        }
    }

    @VisibleForTesting
    internal fun decodeSource(source: LuminanceSource): String? {
        return try {
            val bitmap = createBinaryBitmap(source)
            val rawResult = multiFormatReader.decodeWithState(bitmap)
            if (rawResult != null) {
                qrState = STATE_QRCODE_EXIST
                rawResult.toString()
            } else {
                qrState = STATE_FIND_QRCODE
                null
            }
        } catch (e: Exception) {
            qrState = STATE_FIND_QRCODE
            null
        } finally {
            multiFormatReader.reset()
        }
    }
```

Reply:
Thanks for also testing this. I actually started with different try blocks for each decoding by the `multiFormatReader`, the second one with the inverted source only for the result being null. But with all devices I tested, I did not receive any other exception, so I simplified the code to try and decode the inverted source when the exception is thrown.

Reply:
I have made the changes and also adjusted the tests.